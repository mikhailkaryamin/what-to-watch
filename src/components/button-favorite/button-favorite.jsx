import React from 'react';
import {connect} from 'react-redux';

import {
  func,
  string,
  bool,
} from 'prop-types';
import {filmPropTypes} from '../../types.js';

import {Operation as FavoriteOperation} from '../../reducer/favorite/favorite.js';

const ButtonFavorite = (props) => {
  const {
    film,
    isFavorite,
    place,
    removeFavorite,
    setFavorite,
  } = props;

  const getButtonInFavorite = () => {
    return (
      <button
        className="btn movie-card__button btn--list"
        type="button"
        onClick={() => setFavorite(film, place)}
      >

        <svg width="19" height="20" viewBox="0 0 19 20">
          <use xlinkHref="#add"></use>
        </svg>

        <span>
          My list
        </span>
      </button>
    );
  };

  const getButtonRemoveFavorite = () => {
    return (
      <button
        className="btn movie-card__button btn--list"
        type="button"
        onClick={() => removeFavorite(film, place)}
      >
        <svg width="18" height="14" viewBox="0 0 18 14">
          <use xlinkHref="#in-list"></use>
        </svg>

        <span>
          My list
        </span>
      </button>
    );
  };

  return (
    isFavorite
      ? getButtonRemoveFavorite()
      : getButtonInFavorite()
  );
};

ButtonFavorite.propTypes = {
  film: filmPropTypes,
  isFavorite: bool.isRequired,
  place: string.isRequired,
  setFavorite: func.isRequired,
  removeFavorite: func.isRequired,
};

const mapDispatchToProps = (dispatch) => ({
  removeFavorite: (film, place) => {
    dispatch(FavoriteOperation.removeFavoriteFilm(film, place));
  },
  setFavorite: (film, place) => {
    dispatch(FavoriteOperation.setFavoriteFilm(film, place));
  }
});

export {ButtonFavorite};

export default connect(null, mapDispatchToProps)(ButtonFavorite);
