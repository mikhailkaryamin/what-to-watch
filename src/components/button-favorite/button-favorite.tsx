import * as React from 'react';
import {connect} from 'react-redux';

import {FilmType} from '../../types';

import {Operation as FavoriteOperation} from '../../reducer/favorite/favorite';

type Props = {
  film: FilmType,
  isFavorite: boolean,
  place: string,
  onSetFavorite: (arg0: FilmType, arg1: string) => void,
  onRemoveFavorite: (arg0: FilmType, arg1: string) => void,
}

const ButtonFavorite: React.FC<Props> = (props: Props) => {
  const {
    film,
    isFavorite,
    place,
    onRemoveFavorite,
    onSetFavorite,
  } = props;

  const getButtonInFavorite = () => {
    return (
      <button
        className="btn movie-card__button btn--list"
        type="button"
        onClick={() => onSetFavorite(film, place)}
      >

        <svg width="19" height="20" viewBox="0 0 19 20">
          <use xlinkHref="#add"></use>
        </svg>

        <span>
          My list
        </span>
      </button>
    );
  };

  const getButtonRemoveFavorite = () => {
    return (
      <button
        className="btn movie-card__button btn--list"
        type="button"
        onClick={() => onRemoveFavorite(film, place)}
      >
        <svg width="18" height="14" viewBox="0 0 18 14">
          <use xlinkHref="#in-list"></use>
        </svg>

        <span>
          My list
        </span>
      </button>
    );
  };

  return (
    isFavorite
      ? getButtonRemoveFavorite()
      : getButtonInFavorite()
  );
};

const mapDispatchToProps = (dispatch: (arg0: () => void) => void) => ({
  onRemoveFavorite: (film: FilmType, place: string) => {
    dispatch(FavoriteOperation.removeFavoriteFilm(film, place));
  },
  onSetFavorite: (film: FilmType, place: string) => {
    dispatch(FavoriteOperation.setFavoriteFilm(film, place));
  }
});

export {ButtonFavorite};

export default connect(null, mapDispatchToProps)(ButtonFavorite);
