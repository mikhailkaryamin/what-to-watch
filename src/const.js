const AppRoute = {
  ADD_COMMENT: `/films/:id/review`,
  FILM: `/films/:id`,
  LOGIN: `/login`,
  MY_LIST: `/mylist`,
  PLAYER: `/player/:id`,
  ROOT: `/`,
};

const AuthStatus = {
  AUTH: `AUTH`,
  NO_AUTH: `NO_AUTH`,
};

const ButtonType = {
  ADD_REVIEW: `ADD_REVIEW`,
  PLAY: `play`,
  LIST: `list`,
};

const EscKeyName = {
  LONG: `Escape`,
  SHORT: `Esc`,
};

const FilmCardsListType = {
  MAIN: `main`,
  LIKE_THIS: `like-this`,
};

const FilmDetailedTabsType = {
  OVERVIEW: `Overview`,
  DETAILS: `Details`,
  REVIEWS: `Reviews`,
};

const SpaceKeyName = {
  CODE: 32,
  TEXT: `Space`,
};

export {
  AppRoute,
  AuthStatus,
  ButtonType,
  EscKeyName,
  FilmCardsListType,
  FilmDetailedTabsType,
  SpaceKeyName,
};
