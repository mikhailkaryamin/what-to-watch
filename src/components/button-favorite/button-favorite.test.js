import React from 'react';
import renderer from 'react-test-renderer';

import {FavoriteButtonPlace} from '../../const.js';

import {ButtonFavorite} from './button-favorite.jsx';

describe(`button favorite`, () => {

  test(`should render button remove favorite`, () => {

    const wrapper = renderer
      .create(
          <ButtonFavorite
            isFavorite={true}
            place={FavoriteButtonPlace.DETAILED}
            onRemoveFavorite={()=>{}}
            onSetFavorite={()=>{}}
          />
      ).toJSON();
    expect(wrapper).toMatchSnapshot();
  });

  test(`should render button in favorite`, () => {

    const wrapper = renderer
      .create(
          <ButtonFavorite
            isFavorite={false}
            place={FavoriteButtonPlace.DETAILED}
            onRemoveFavorite={()=>{}}
            onSetFavorite={()=>{}}
          />
      ).toJSON();
    expect(wrapper).toMatchSnapshot();
  });
});

