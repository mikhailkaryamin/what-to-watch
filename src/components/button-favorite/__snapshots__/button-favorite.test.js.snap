// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`button favorite should render button favorite 1`] = `
<button
  className="btn movie-card__button btn--list"
  onClick={[Function]}
  type="button"
>
  <svg
    height="14"
    viewBox="0 0 18 14"
    width="18"
  >
    <use
      xlinkHref="#in-list"
    />
  </svg>
  <span>
    My list
  </span>
</button>
`;
